File: DSA/done/36 - isValidSudoku.py
# https://leetcode.com/problems/valid-sudoku/

from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = {}
        rows = {}
        squares = {}
        
        for r in range(9):
            rows[r] = set()
            for c in range(9):
                if board[r][c] == ".":
                    continue

                if c not in cols:
                    cols[c] = set()
                if (r // 3, c // 3) not in squares:
                    squares[(r // 3, c // 3)] = set()
                
                if (board[r][c] in rows[r]
                    or board[r][c] in cols[c]
                    or board[r][c] in squares[(r // 3, c // 3)]):
                    return False
                
                cols[c].add(board[r][c])
                rows[r].add(board[r][c])
                squares[(r // 3, c // 3)].add(board[r][c])
        
        return True

File: DSA/done/36 - isValidSudoku.cpp
// https://leetcode.com/problems/valid-sudoku/

#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        unordered_map<int, unordered_set<char>> rows, cols, squares;

        for (int r{0}; r < 9; r++) {
            for (int c{0}; c < 9; c++) {
                if (board[r][c] == '.') continue;

                int squareKey = (r / 3) * 3 + (c / 3);  // encodes to 0–8

                if (
                    rows[r].count(board[r][c]) ||
                    cols[c].count(board[r][c]) ||
                    squares[squareKey].count(board[r][c])
                )
                    return false;

                rows[r].insert(board[r][c]);
                cols[c].insert(board[r][c]);
                squares[squareKey].insert(board[r][c]);
            }
        }
        return true;
    }
};
